File: apps/shell/electron/coordinator/regenerate.ts
// Per-item regeneration. User rejects / dislikes an item → we re-run
// agent-3 (designer) for JUST that slot, without showing the reviewer's
// concerns to the designer (D1.b — keep agent boundaries clean, rely on
// slot_spec + KU as the only inputs).
//
// Mechanism: temporarily reduce blackboard.mapping.blueprint.slot_specs
// to the single target slot + clear data.items, run /agent-3-designer,
// read back the new item, then restore original state with the new item
// swapped in. The reviewer re-run is NOT automatic; user re-triggers
// "Run Gemini Second Opinion" (or a future "Re-review" flow) to get
// fresh verdicts on the regenerated items.

import path from 'node:path';
import { readBlackboard, writeBlackboard } from './blackboard';
import { coordinatorEvents, spawnAgent } from './workflow';
import type { Blackboard } from './types';

// ─── types (loose; mirrors blackboard schema) ──────────────

interface BBItem {
  item_id?: string;
  slot_index?: number;
  [key: string]: unknown;
}

interface BBSlotSpec {
  slot_index?: number;
  [key: string]: unknown;
}

interface BBMapping {
  blueprint?: { slot_specs?: BBSlotSpec[]; total_slots?: number; [k: string]: unknown };
  [key: string]: unknown;
}

interface BBReview {
  per_item?: Array<{ item_id: string; [k: string]: unknown }>;
  summary?: Record<string, unknown>;
}

// ─── guardrails ─────────────────────────────────────────────

let currentItemId: string | null = null;
let batchRunning = false;

export function isRegenerating(): boolean {
  return currentItemId !== null || batchRunning;
}

export function currentRegenerationItem(): string | null {
  return currentItemId;
}

// ─── main entrypoint ────────────────────────────────────────

export async function regenerateItem(workspaceDir: string, itemId: string): Promise<void> {
  if (currentItemId) throw new Error('another item is already being regenerated');
  const blackboardPath = path.join(workspaceDir, 'blackboard.json');

  const original = await readBlackboard(blackboardPath);
  if (!original) throw new Error('blackboard.json not readable');

  const originalItems = (original.data.items ?? []) as BBItem[];
  const itemIdx = originalItems.findIndex((i) => i.item_id === itemId);
  if (itemIdx < 0) throw new Error(`item ${itemId} not found`);
  const targetSlotIndex = originalItems[itemIdx].slot_index;
  if (typeof targetSlotIndex !== 'number') {
    throw new Error(`item ${itemId} has no slot_index`);
  }

  const originalMapping = original.data.mapping as BBMapping | null;
  const slotSpecs = originalMapping?.blueprint?.slot_specs ?? [];
  const slotSpec = slotSpecs.find((s) => s.slot_index === targetSlotIndex);
  if (!slotSpec) throw new Error(`slot_spec for slot_index ${targetSlotIndex} not found`);

  currentItemId = itemId;
  coordinatorEvents.emit('regenerate:started', { item_id: itemId });

  try {
    // Build reduced blackboard: same user_input, same KUs, but only ONE
    // slot in blueprint + no existing items. agent-3 will produce one
    // item — we'll splice it back into originalItems.
    const reduced: Blackboard = {
      ...original,
      workflow: { ...original.workflow, current_step: 2, status: 'pending' },
      data: {
        ...original.data,
        items: null,
        review: null,
        // Preserve other review blocks verbatim — agent-3 doesn't read them,
        // and we'll restore/filter after the run anyway.
        mapping: {
          ...(originalMapping ?? {}),
          blueprint: {
            ...(originalMapping?.blueprint ?? {}),
            total_slots: 1,
            slot_specs: [slotSpec],
          },
        } as unknown as Blackboard['data']['mapping'],
      },
    };

    await writeBlackboard(blackboardPath, reduced);
    const { result } = await spawnAgent('agent_3', workspaceDir);

    // Read back the result; agent-3 should have written exactly one item
    const post = await readBlackboard(blackboardPath);
    if (!post) throw new Error('blackboard missing after agent-3');
    const newItems = (post.data.items ?? []) as BBItem[];
    if (newItems.length === 0) throw new Error('agent-3 produced no item');
    const newItem = newItems[0];
    // Force the original item_id + slot_index so downstream code can find it
    newItem.item_id = itemId;
    newItem.slot_index = targetSlotIndex;
    // Clear any stale user_override (post-regen the item is new, unjudged)
    delete newItem.user_override;

    // Restore original blackboard + splice the new item in place
    const restored: Blackboard = { ...original };
    restored.data = {
      ...original.data,
      items: originalItems.map((it, i) => (i === itemIdx ? newItem : it)) as unknown[],
      mapping: originalMapping as Blackboard['data']['mapping'],
      review: null,
    };
    restored.workflow = { ...original.workflow };

    // Drop stale review entries for the regenerated item across all three
    // review blocks. Summary fields stay stale until user re-runs review.
    const dropByItemId = (r: BBReview | null | undefined): BBReview | null => {
      if (!r) return r ?? null;
      const per = Array.isArray(r.per_item) ? r.per_item.filter((e) => e.item_id !== itemId) : r.per_item;
      return { ...r, per_item: per };
    };
    restored.data.review_claude = dropByItemId(restored.data.review_claude as BBReview | null) as Blackboard['data']['review_claude'];
    restored.data.review_gemini = dropByItemId(restored.data.review_gemini as BBReview | null) as Blackboard['data']['review_gemini'];
    restored.data.review_merged = dropByItemId(restored.data.review_merged as BBReview | null) as Blackboard['data']['review_merged'];

    // Accrue regenerate cost into a dedicated bucket so the main
    // per-agent cost totals stay meaningful.
    restored.costs = restored.costs ?? { total_usd: 0, by_agent: {} };
    const prior = (restored.costs.by_agent.regenerate as number) ?? 0;
    restored.costs.by_agent.regenerate = prior + (result?.total_cost_usd ?? 0);
    restored.costs.total_usd = Object.values(restored.costs.by_agent).reduce<number>(
      (s, v) => s + (typeof v === 'number' ? v : 0),
      0,
    );

    await writeBlackboard(blackboardPath, restored);
    coordinatorEvents.emit('board:updated', { board: restored });
    coordinatorEvents.emit('regenerate:completed', {
      item_id: itemId,
      cost_usd: result?.total_cost_usd ?? 0,
      duration_ms: result?.duration_ms ?? 0,
    });
  } catch (err) {
    coordinatorEvents.emit('regenerate:error', { item_id: itemId, error: (err as Error).message });
    throw err;
  } finally {
    currentItemId = null;
  }
}

// ─── batch: re-run all items with user_override='reject' ────

export async function regenerateRejected(workspaceDir: string): Promise<void> {
  if (batchRunning) throw new Error('batch already running');
  const blackboardPath = path.join(workspaceDir, 'blackboard.json');
  const snap = await readBlackboard(blackboardPath);
  if (!snap) throw new Error('blackboard.json not readable');
  const items = (snap.data.items ?? []) as BBItem[];
  const rejectedIds = items
    .filter((it) => (it.user_override as string | undefined) === 'reject')
    .map((it) => it.item_id as string)
    .filter(Boolean);

  if (rejectedIds.length === 0) {
    coordinatorEvents.emit('regenerate-batch:completed', { count: 0, regenerated: [] });
    return;
  }

  batchRunning = true;
  coordinatorEvents.emit('regenerate-batch:started', { item_ids: rejectedIds });
  const completed: string[] = [];
  try {
    for (const id of rejectedIds) {
      try {
        await regenerateItem(workspaceDir, id);
        completed.push(id);
        coordinatorEvents.emit('regenerate-batch:item-done', {
          item_id: id,
          remaining: rejectedIds.length - completed.length,
        });
      } catch (err) {
        // One failure shouldn't abort the whole batch; record + continue.
        coordinatorEvents.emit('regenerate:error', {
          item_id: id,
          error: (err as Error).message,
        });
      }
    }
    coordinatorEvents.emit('regenerate-batch:completed', {
      count: completed.length,
      regenerated: completed,
    });
  } finally {
    batchRunning = false;
  }
}

File: apps/quiz/src-tauri/src/regenerate.rs
// Per-item regeneration — Rust port of
// apps/shell/electron/coordinator/regenerate.ts.
//
// User rejects an item → re-run agent-3 for ONE slot only, then splice
// the new item back into the original blackboard. Reviewers don't re-run
// here; user re-triggers Gemini second-opinion separately.

use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use serde_json::{json, Value};
use tauri::{AppHandle, Emitter};
use tokio::sync::Mutex;

use crate::blackboard::{read_blackboard, write_blackboard};
use crate::types::Blackboard;
use crate::workflow::{spawn_agent, WorkflowRuntime};

#[derive(Default)]
pub struct RegenerateRuntime {
    pub current_item_id: Mutex<Option<String>>,
    pub batch_running: AtomicBool,
}

impl RegenerateRuntime {
    pub async fn is_busy(&self) -> bool {
        self.current_item_id.lock().await.is_some()
            || self.batch_running.load(Ordering::SeqCst)
    }
}

// ─── single-item regeneration ───────────────────────────────

pub async fn regenerate_item(
    app: AppHandle,
    workspace_dir: PathBuf,
    item_id: String,
    workflow_runtime: Arc<WorkflowRuntime>,
    regen_runtime: Arc<RegenerateRuntime>,
) {
    let result = regenerate_item_inner(
        &app,
        &workspace_dir,
        &item_id,
        &workflow_runtime,
        &regen_runtime,
    )
    .await;
    if let Err(e) = result {
        let _ = app.emit(
            "regenerate:error",
            json!({ "item_id": item_id, "error": e }),
        );
    }
}

async fn regenerate_item_inner(
    app: &AppHandle,
    workspace_dir: &Path,
    item_id: &str,
    workflow_runtime: &Arc<WorkflowRuntime>,
    regen_runtime: &Arc<RegenerateRuntime>,
) -> Result<(), String> {
    {
        let mut g = regen_runtime.current_item_id.lock().await;
        if g.is_some() {
            return Err("another item is already being regenerated".into());
        }
        *g = Some(item_id.to_string());
    }
    let blackboard_path = workspace_dir.join("blackboard.json");

    let result = run_regenerate(
        app,
        &blackboard_path,
        workspace_dir,
        item_id,
        workflow_runtime,
    )
    .await;

    {
        let mut g = regen_runtime.current_item_id.lock().await;
        *g = None;
    }
    result
}

async fn run_regenerate(
    app: &AppHandle,
    blackboard_path: &Path,
    workspace_dir: &Path,
    item_id: &str,
    workflow_runtime: &Arc<WorkflowRuntime>,
) -> Result<(), String> {
    let original = read_blackboard(blackboard_path)
        .await
        .ok_or_else(|| "blackboard.json not readable".to_string())?;
    let original_items = original
        .data
        .items
        .clone()
        .ok_or_else(|| "no items present".to_string())?;
    let item_idx = original_items
        .iter()
        .position(|it| it.get("item_id").and_then(|v| v.as_str()) == Some(item_id))
        .ok_or_else(|| format!("item {item_id} not found"))?;
    let target_slot_index = original_items[item_idx]
        .get("slot_index")
        .and_then(|v| v.as_i64())
        .ok_or_else(|| format!("item {item_id} has no slot_index"))?;

    let original_mapping = original.data.mapping.clone().unwrap_or_default();
    let original_blueprint = original_mapping
        .get("blueprint")
        .cloned()
        .unwrap_or_else(|| Value::Object(Default::default()));
    let slot_spec = original_blueprint
        .get("slot_specs")
        .and_then(|v| v.as_array())
        .and_then(|specs| {
            specs.iter().find(|s| {
                s.get("slot_index").and_then(|v| v.as_i64()) == Some(target_slot_index)
            })
        })
        .cloned()
        .ok_or_else(|| format!("slot_spec for slot_index {target_slot_index} not found"))?;

    let _ = app.emit("regenerate:started", json!({ "item_id": item_id }));

    // Build reduced blackboard: same user_input + KUs but only ONE slot.
    let mut reduced: Blackboard = original.clone();
    reduced.workflow.current_step = 2;
    reduced.workflow.status = crate::types::WorkflowStatus::Pending;
    reduced.data.items = None;
    reduced.data.review = None;

    let mut reduced_blueprint = original_blueprint.clone();
    if let Value::Object(map) = &mut reduced_blueprint {
        map.insert("total_slots".into(), json!(1));
        map.insert("slot_specs".into(), json!([slot_spec]));
    }
    let mut reduced_mapping = original_mapping.clone();
    reduced_mapping.insert("blueprint".into(), reduced_blueprint);
    reduced.data.mapping = Some(reduced_mapping);

    write_blackboard(blackboard_path, &reduced).await?;

    let result = spawn_agent(app, "agent_3", workspace_dir, workflow_runtime).await?;

    // Read back; expect exactly one new item.
    let post = read_blackboard(blackboard_path)
        .await
        .ok_or_else(|| "blackboard missing after agent-3".to_string())?;
    let new_items = post.data.items.unwrap_or_default();
    if new_items.is_empty() {
        return Err("agent-3 produced no item".into());
    }
    let mut new_item = new_items.into_iter().next().unwrap();
    if let Value::Object(map) = &mut new_item {
        map.insert("item_id".into(), json!(item_id));
        map.insert("slot_index".into(), json!(target_slot_index));
        map.remove("user_override");
    }

    // Restore original blackboard with the new item swapped in.
    let mut restored: Blackboard = original.clone();
    let mut items: Vec<Value> = original_items.clone();
    items[item_idx] = new_item;
    restored.data.items = Some(items);
    restored.data.mapping = original.data.mapping.clone();
    restored.data.review = None;

    // Drop stale review entries for the regenerated item.
    drop_review_entry(&mut restored.data.review_claude, item_id);
    drop_review_entry(&mut restored.data.review_gemini, item_id);
    drop_review_entry(&mut restored.data.review_merged, item_id);

    // Cost: accrue into a dedicated "regenerate" bucket.
    let prior = restored
        .costs
        .by_agent
        .get("regenerate")
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);
    let added = result.as_ref().and_then(|r| r.total_cost_usd).unwrap_or(0.0);
    restored
        .costs
        .by_agent
        .insert("regenerate".to_string(), json!(prior + added));
    restored.costs.total_usd = restored
        .costs
        .by_agent
        .values()
        .filter_map(|v| v.as_f64())
        .sum();

    write_blackboard(blackboard_path, &restored).await?;
    let _ = app.emit("board:updated", json!({ "board": &restored }));
    let _ = app.emit(
        "regenerate:completed",
        json!({
            "item_id": item_id,
            "cost_usd": added,
            "duration_ms": result.as_ref().and_then(|r| r.duration_ms).unwrap_or(0),
        }),
    );
    Ok(())
}

fn drop_review_entry(
    block: &mut Option<std::collections::HashMap<String, Value>>,
    item_id: &str,
) {
    let Some(map) = block else { return };
    if let Some(per_item) = map.get_mut("per_item") {
        if let Some(arr) = per_item.as_array_mut() {
            arr.retain(|x| x.get("item_id").and_then(|v| v.as_str()) != Some(item_id));
        }
    }
}

// ─── batch regeneration of all rejected items ──────────────

pub async fn regenerate_rejected(
    app: AppHandle,
    workspace_dir: PathBuf,
    workflow_runtime: Arc<WorkflowRuntime>,
    regen_runtime: Arc<RegenerateRuntime>,
) {
    if regen_runtime.batch_running.swap(true, Ordering::SeqCst) {
        let _ = app.emit(
            "regenerate:error",
            json!({ "item_id": "", "error": "batch already running" }),
        );
        return;
    }
    let result = run_batch(&app, &workspace_dir, &workflow_runtime, &regen_runtime).await;
    regen_runtime.batch_running.store(false, Ordering::SeqCst);
    if let Err(e) = result {
        let _ = app.emit(
            "regenerate:error",
            json!({ "item_id": "", "error": e }),
        );
    }
}

async fn run_batch(
    app: &AppHandle,
    workspace_dir: &Path,
    workflow_runtime: &Arc<WorkflowRuntime>,
    regen_runtime: &Arc<RegenerateRuntime>,
) -> Result<(), String> {
    let blackboard_path = workspace_dir.join("blackboard.json");
    let snap = read_blackboard(&blackboard_path)
        .await
        .ok_or_else(|| "blackboard.json not readable".to_string())?;
    let items = snap.data.items.unwrap_or_default();
    let rejected_ids: Vec<String> = items
        .iter()
        .filter(|it| {
            it.get("user_override").and_then(|v| v.as_str()) == Some("reject")
        })
        .filter_map(|it| {
            it.get("item_id").and_then(|v| v.as_str()).map(|s| s.to_string())
        })
        .collect();

    if rejected_ids.is_empty() {
        let _ = app.emit(
            "regenerate-batch:completed",
            json!({ "count": 0, "regenerated": [] }),
        );
        return Ok(());
    }

    let _ = app.emit(
        "regenerate-batch:started",
        json!({ "item_ids": &rejected_ids }),
    );

    let mut completed: Vec<String> = Vec::new();
    let total = rejected_ids.len();
    for id in rejected_ids {
        match regenerate_item_inner(
            app,
            workspace_dir,
            &id,
            workflow_runtime,
            regen_runtime,
        )
        .await
        {
            Ok(()) => {
                completed.push(id.clone());
                let _ = app.emit(
                    "regenerate-batch:item-done",
                    json!({
                        "item_id": id,
                        "remaining": total - completed.len(),
                    }),
                );
            }
            Err(e) => {
                let _ = app.emit(
                    "regenerate:error",
                    json!({ "item_id": id, "error": e }),
                );
            }
        }
    }

    let _ = app.emit(
        "regenerate-batch:completed",
        json!({
            "count": completed.len(),
            "regenerated": completed,
        }),
    );
    Ok(())
}
